contracts/utils/OptionsPurchaseQueue.sol

```diff
@@ -0,0 +1,442 @@
+     */
+    function requestPurchase(address vault, uint256 optionsAmount)
+        external
+        override
+        returns (uint256)
+    {
+        require(whitelistedBuyer[msg.sender], "Buyer not whitelisted");
+        uint256 _ceilingPrice = ceilingPrice[vault];
+        require(_ceilingPrice != 0, "Invalid vault");
+        // This prevents new purchase requested after the vault has set its allocation
+        require(vaultAllocatedOptions[msg.sender] == 0, "Vault allocated");
+        require(optionsAmount != 0, "!optionsAmount");
+
+        // premiums = optionsAmount * ceilingPrice
+        uint256 premiums =
+            optionsAmount.mul(_ceilingPrice).div(10**Vault.OTOKEN_DECIMALS);
+
+        // Add purchase to queue
+        purchases[vault].push(
+            Purchase(
+                SafeCast.toUint128(optionsAmount),
+                SafeCast.toUint128(premiums),
+                msg.sender
+            )
+        );
+
+        totalOptionsAmount[vault] = totalOptionsAmount[vault].add(
+            optionsAmount
+        );
+
+        // Transfer premiums from the buyer to this contract
+        IERC20(IRibbonThetaVault(vault).vaultParams().asset).safeTransferFrom(
+            msg.sender,
+            address(this),
+            premiums
+        );
+
+        emit PurchaseRequested(msg.sender, vault, optionsAmount, premiums);
+
+        return premiums;
+    }
+
+    /**
+     * @notice Allocate options to the purchase queue
+     * @dev Only callable by the vault selling options. Since we cannot allocate more options than there are on the
+     *  purchase queue, we cap the allocated options at the totalOptionsAmount. The vault decides how many options
+     *  of its options it wants to allocate.
+     * @param allocatedOptions Maximum amount of options the vault can allocate to buyers
+     * @return allocatedOptions The actual amount of options allocated
+     */
+    function allocateOptions(uint256 allocatedOptions)
+        external
+        override
+        returns (uint256)
+    {
+        require(ceilingPrice[msg.sender] != 0, "Not vault");
+
+        // Prevent the vault from allocating more options than there are requested
+        uint256 optionsAmount = totalOptionsAmount[msg.sender];
+        // allocatedOptions = min(allocatedOptions, optionsAmount)
+        allocatedOptions = optionsAmount < allocatedOptions
+            ? optionsAmount
+            : allocatedOptions;
+
+        // Blocks new purchase requests until sellToBuyers() is called
+        vaultAllocatedOptions[msg.sender] = allocatedOptions;
+
+        if (allocatedOptions != 0) {
+            // Transfer allocated options from the vault to this contract
+            IERC20(IRibbonThetaVault(msg.sender).currentOption())
+                .safeTransferFrom(msg.sender, address(this), allocatedOptions);
+        }
+
+        emit OptionsAllocated(msg.sender, allocatedOptions);
```

Comment:
Should the emit should go under the `if` branch?

Reply:
I think we can put it put it under since there may be a situation where the vault allocates 0 options to the queue